### PR TITLE
Verify downloaded test runner zip file 812

### LINKED ISSUE #812: Check downloaded binary file size
When downloading binary, it might be corrupted / intercepted by a proxy. Sometimes even CDN has a hiccup like shown below from a test project run
```
Instead we will install version: https://cdn.cypress.io/beta/binary/1.0.3/win64/appveyor-develop-7b4c223233ca8eb6e147dc5e89b30a01061f6581-11678682/cypress.zip
Note: there is no guarantee these versions will work properly together.
Installing Cypress (version: https://cdn.cypress.io/beta/binary/1.0.3/win64/appveyor-develop-7b4c223233ca8eb6e147dc5e89b30a01061f6581-11678682/cypress.zip)
[?25l[18:21:09]  Downloading Cypress     [started]
[18:21:16]  Downloading Cypress     [completed]
[18:21:16]  Unzipping Cypress       [started]
[18:22:16]  Unzipping Cypress       [completed]
[18:22:16]  Finishing Installation  [started]
[18:22:16]  Finishing Installation  [completed]
[?25h
You can now open Cypress by running: node_modules\.bin\cypress open
https://on.cypress.io/installing-cypress
[?25h
> node-sass@4.5.3 postinstall C:\projects\cypress-test-example-repos\cypress-example-piechopper\node_modules\node-sass
> node scripts/build.js
Binary found at C:\projects\cypress-test-example-repos\cypress-example-piechopper\node_modules\node-sass\vendor\win32-ia32-57\binding.node
Testing binary
Binary is fine
npm ERR! path C:\projects\cypress-test-example-repos\cypress-example-piechopper\node_modules\fsevents\node_modules
npm ERR! code EPERM
npm ERR! errno -4048
npm ERR! syscall lstat
```

Need to validate downloaded zip file size. If it < 50 MB something went terribly wrong.

### LINKED ISSUE #3515: Unzipping Cypress fails on Concourse
<!-- Is this a question? Don't open an issue. Ask in our chat https://gitter.im/cypress-io/cypress-->

### Current behavior:

Unzipping Cypress fails on concourse with an error from the unzipping tool used by Cypress, `yauzl`:

```
error /tmp/build/ec0e34e0/src/node_modules/cypress: Command failed.                                                                                                                                                                                                                                                                                                                                                                                                                                                
Exit code: 1
Command: node index.js --exec install
Arguments: 
Directory: /tmp/build/ec0e34e0/src/node_modules/cypress
Output:
Installing Cypress (version: 3.1.0)

25l[16:53:03]  Downloading Cypress     [started]
[16:53:03]  Downloading Cypress      0% 0s [title changed]
[16:53:03]  Downloading Cypress      22% 1s [title changed]
[16:53:04]  Downloading Cypress      95% 0s [title changed]
[16:53:04]  Downloaded Cypress      [title changed]
[16:53:04]  Downloaded Cypress      [completed]
[16:53:04]  Unzipping Cypress       [started]
25hevents.js:183
      throw er; // Unhandled 'error' event
      ^

Error: invalid central directory file header signature: 0x1e02014b
    at /tmp/build/ec0e34e0/src/node_modules/yauzl/index.js:249:70
    at /tmp/build/ec0e34e0/src/node_modules/yauzl/index.js:618:5
    at /tmp/build/ec0e34e0/src/node_modules/fd-slicer/index.js:32:7
    at FSReqWrap.wrapper [as oncomplete] (fs.js:658:17)
```

There also seems to be a previous error when trying to run cypress, not sure if they are related.
<!-- images, stack traces, etc -->

### Desired behavior:

Downloading and running of Cypress successful on Concourse

<!-- A clear description of what you want to happen -->

### Steps to reproduce: (app code and test code)

<!-- Issues without reproducible steps will get closed. You can fork https://github.com/cypress-io/cypress-test-tiny repo, set up a failing test, then tell us the repo/branch to try. -->

Install Cypress on Concourse

### Versions

Cypress 3.1.5
Yarn 1.9.4
Node 8.12.0
<!-- Cypress, operating system, browser -->

### PR DESCRIPTION
**Pull request by bahmutov**
- uses custom meta headers from CDN to verify the downloaded file. If the checksum is available, uses both checksum and file size. If only file size is available, which is content-length, uses file size. Otherwise no verification, but it is hard to think of the case when there is no information
- closes #812
- closes #3515 
- [x] add a test where download goes through, but the checksum is different
- [x] add a test where download goes through, there is no checksum, but file size is different

### CODE REVIEW

**Comment by Bkucera** [comment]
@bahmutov 
Could you point me to where you got those header names? I found this https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonRequestHeaders.html

which says the hash header name is `x-amz-content-sha256`

**Comment by bahmutov** [comment]
Hehehe no these are custom meta tags we set on the S3 resource ourselves!

**Comment by Bkucera** [comment]
oh, nice. And its a sha-512 ?

**Comment by bahmutov** [comment]
256

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
All dependencies within the `cli` package should support Node 4+, `hasha`, after 3.0.0, does not support Node 4, I'd recommend downgrading this package to 3.0.0 or require a new package that supports Node 4.

**Comment by bahmutov** [comment]
Hmm we should drop node 4 then - even node 6 is no longer supported by node itself

**Comment by jennifer-shehane** [comment]
Yah, probably a discussion we need to have.

**Comment by Bkucera** [comment]
Hasha defaults to sha-512 according to readme https://github.com/sindresorhus/hasha/blob/master/readme.md#algorithm

**Comment by bahmutov** [comment]
Hmm maybe i misremember it then since it uses the default setting, 512 is good

**Comment by bahmutov** [comment]
I moved hasha to be dev dependency of CLI - and lifted the little code snippet for computing sha512 hash of a file using just system Node dependencies

**Review by kuceb — DISMISSED** [review]

**Comment by bahmutov** [comment]
Since I removed Hasha as a production dependency in CLI, it is no longer changing Node 4 support

**Review by jennifer-shehane — DISMISSED** [review]

**Comment by bahmutov** [comment]
verifying the downloaded file should resolve unzipping issues reported in #3515

File: cli/package.json [hunk 1]
```diff
@@ -92,6 +92,7 @@
     "dependency-check": "3.3.0",
     "dtslint": "0.7.1",
     "execa-wrap": "1.4.0",
+    "hasha": "5.0.0",
```
Comment (brian-mann) [inline cli/package.json]: is this intentional

Comment (bahmutov) [reply cli/package.json]: yup, code was lifted, but dev dependency to validate inside a test

File: cli/lib/tasks/download.js [hunk 2]
```diff
@@ -79,6 +79,95 @@ const prettyDownloadErr = (err, version) => {
...
+ * values or just one of them to be checked.
+ */
+const verifyDownloadedFile = (filename, expectedSize, expectedChecksum) => {
+  if (expectedSize && expectedChecksum) {
+    debug('verifying checksum and file size')
+
+    return Promise.join(
+      util.getFileChecksum(filename),
+      util.getFileSize(filename)
+    ).spread((checksum, filesize) => {
```
Comment (pietrofxq) [inline cli/lib/tasks/download.js]: I think Promise.join without the function handler parameter is deprecated, per https://github.com/petkaantonov/bluebird/issues/1482

Comment (bahmutov) [reply cli/lib/tasks/download.js]: thank you, I will update it to not use deprecated form

**Comment by bfagundez** [comment]
Hello, seems like the test failing is in fact the same problem we are having [here](https://github.com/cypress-io/cypress/pull/4193). 
Is this going to be looked into anytime soon?

**Comment by jennifer-shehane** [comment]
@bahmutov This test is consistently failing on your PR, can you take a look please?

File: cli/test/lib/tasks/download_spec.js [hunk 3]
```diff
@@ -135,6 +136,103 @@ describe('lib/tasks/download', function () {
...
+    })
+
+    it('throws if checksum and file size are different from expected', function () {
+      nock('https://download.cypress.io')
+      .get('/desktop/1.2.3')
+      .query(true)
+      .reply(200, () => {
+        return fs.createReadStream('test/fixture/example.zip')
+      }, {
+        'x-amz-meta-checksum': 'incorrect-checksum',
```
Comment (flotwig) [inline cli/test/lib/tasks/download_spec.js]: neat, didn't know you could add headers to s3 resources like this

**Review by flotwig — DISMISSED** [review]
LGTM once failing test is fixed

**Review by flotwig — APPROVED** [review]

**Pull request merged**